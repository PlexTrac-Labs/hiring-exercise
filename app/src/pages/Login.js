import React, { useRef, useState, useEffect } from "react";
import axios from "axios";
import CredentialWarning from "../components/CredentialWarning";

const LOCAL_STORAGE_KEY = "app.user";

export default function Login() {
  const BASE_URL = "http://localhost:5000"; //todo figure out how to have this just reference path

  //Need a state to store userId
  const [userId, setUserId] = useState([]);
  const [isWrongLogin, setIsWrongLogin] = useState(false);

  //References to Password and Name in form
  const nameRef = useRef();
  const passwordRef = useRef();

  //Add userId to cookie once known, so we can use it in rest of site.
  useEffect(() => {
    localStorage.setItem(LOCAL_STORAGE_KEY, JSON.stringify(userId));
  }, [userId]);

  function handleLoginSubmit(e) {
    e.preventDefault(); // Prevent html from sending the request itself
    const name = nameRef.current.value;
    const password = passwordRef.current.value;
    if (name === "" || password === "") return;

    //Lets try this out, see how clean it is
    axios
      .post(`${BASE_URL}/authenticate`, {
        username: name,
        password: password
      })
      .then(response => {
        setUserId(response.data);
        window.location = "/profile";
        console.log(response.data);
      })
      .catch(error => {
        setIsWrongLogin(true);
      });
  }

  return (
    <div>
      <form className="login" onClick={handleLoginSubmit}>
        <label>
          Name:
          <input ref={nameRef} type="text" name="name" required />
        </label>
        <br />
        <label>
          Password:
          <input ref={passwordRef} type="text" name="password" required />
        </label>
        <br />
        <input type="submit" />
        {isWrongLogin ? <CredentialWarning /> : null}
      </form>
    </div>
  );
}
