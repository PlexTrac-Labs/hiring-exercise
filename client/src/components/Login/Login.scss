.login-wrapper {
  color: white;
  text-align: center;

  .login-form {
    display: flex;
    flex-direction: column;
    align-items: center;
    margin: 16px;

    > * {
      margin-top: 8px;
    }

    .username-input {
      color: white;
      width: 250px;
      text-align: start;
      input {
        color: white;
      }
    }
    .password-input {
      color: white;
      width: 250px;
      input {
        color: white;
      }
    }
    .login-btn {
      color: white;
    }
  }
}
