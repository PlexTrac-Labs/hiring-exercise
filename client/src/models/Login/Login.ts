import React from "react";

export interface ILogin {
  Username: string;
  Password: string;
}
