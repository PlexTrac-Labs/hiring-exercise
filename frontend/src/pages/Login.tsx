import {
  Alert,
  Box,
  Button,
  FormControl,
  Grid,
  TextField,
  Typography,
} from "@mui/material";
import { styled } from "@mui/system";
import React, { useState } from "react";
import { useNavigate } from "react-router-dom";
import { AUTH_TOKEN } from "../helpful/constants";
import { authenticate } from "../util/user/user";

const LoginBox = styled(Box)(({ theme }) => ({
  display: "flex",
  flexDirection: "column",
  padding: 15,
  //backgroundColor: "gray",
  justifyContent: "center",
  alignItems: "center",
  //height: "100vh",
}));

export const Login = () => {
  const navigate = useNavigate();
  const [error, setError] = useState(false);

  const handleSubmit = async (event: any) => {
    event.preventDefault();
    const username = event.target.username.value;
    const password = event.target.password.value;
    try {
      const auth = await authenticate(username, password);
      localStorage.setItem(AUTH_TOKEN, auth.auth_token);
      navigate("list");
    } catch (error) {
      setError(true);
    }
  };

  return (
    <LoginBox>
      <form onSubmit={handleSubmit}>
        <Grid
          container
          direction="column"
          rowSpacing={2}
          alignItems="center"
          justifyContent="center"
        >
          <Grid item xs={12}>
            <Typography variant="h3">Login</Typography>
          </Grid>
          {error && <Alert severity="error">Invalid Username/Password</Alert>}

          <Grid item xs={12}>
            <FormControl>
              <TextField
                required
                id="username"
                variant="outlined"
                label="Username"
              />
            </FormControl>
          </Grid>
          <Grid item xs={12}>
            <FormControl>
              <TextField
                required
                id="password"
                variant="outlined"
                label="Password"
                type="password"
              />
            </FormControl>
          </Grid>
          <Grid item xs={12}>
            <Button color="primary" variant="contained" type="submit">
              Login
            </Button>
          </Grid>
        </Grid>
      </form>
    </LoginBox>
  );
};
