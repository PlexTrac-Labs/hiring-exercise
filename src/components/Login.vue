<template>
  <div class="row">
    <h1 class="col-12">Please Login</h1>
  </div>
  <div class="row mt-4">
    <div class="col-6 offset-3">
      <form v-on:submit="submit" class="border border-dark p-5">
        <div class="form-group form-horizontal control-label text-left">
          <label for="userNameInput">Username</label>
          <input
            class="form-control bg-light"
            id="userNameInput"
            aria-describedby="emailHelp"
            placeholder="Enter username"
            required
            v-model="username"
          />
        </div>
        <div class="form-group form-horizontal control-label text-left">
          <label for="passwordInput">Password</label>
          <input
            type="password"
            class="form-control bg-light"
            id="passwordInput"
            placeholder="Enter password"
            required
            v-model="password"
          />
        </div>
        <button type="submit" class="btn btn-primary">Submit</button>
      </form>
    </div>
  </div>
</template>
<script>
export default {
  data: function() {
    return {
      username: "",
      password: ""
    };
  },
  methods: {
    submit(event) {
      event.preventDefault();
      let payload = {
        username: this.username,
        password: this.password
      };
      this.$store.dispatch("authenticate", payload);
    }
  }
};
</script>
<style scoped>
label {
  text-align: left !important;
}
</style>
