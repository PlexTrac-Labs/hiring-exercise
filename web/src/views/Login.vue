<template>
  <v-container>
    <v-row>
      <v-col cols="4">
        <v-text-field
          v-model="username"
          placeholder="< Username >"
          :error-messages="errorMessages"
          :loading="loading"
        />
      </v-col>
    </v-row>
    <v-row>
      <v-col cols="4">
        <v-text-field
          v-model="password"
          placeholder="< Password >"
          :error-messages="errorMessages"
          :loading="loading"
          :type="showPassword ? 'text' : 'password'"
          @keyup.enter.native="connect(password)"
        >
          <v-icon
            v-if="showPassword"
            slot="append"
            color="grey"
            @click="toggleShowPassword"
          >
            far fa-eye-slash
          </v-icon>
          <v-icon
            v-if="!showPassword"
            slot="append"
            color="grey"
            @click="toggleShowPassword"
          >
            far fa-eye
          </v-icon>
        </v-text-field>
      </v-col>
    </v-row>
    <v-row>
      <v-btn
        color="info"
        :disabled="!Boolean(password && username)"
        @click="login(username, password)"
      >
        login
      </v-btn>
    </v-row>
  </v-container>
</template>

<script>
export default {
  name: "Login",
  components: {},
  data: () => ({
    password: "",
    username: "",
    errorMessages: [],
    showPassword: false,
    loading: false
  }),
  methods: {
    toggleShowPassword() {
      this.showPassword = !this.showPassword;
    },
    async login(username, password) {
      console.log("logging in with: ", username, password);
      try {
        this.loading = true;
        // TODO: keep track of the returned auth_token for future requests
        await this.$http({
          method: "post",
          url: "http://localhost:5000/authenticate",
          data: {
            username,
            password // this is 100% not secure (with http) :) TODO: stop this?
          },
          headers: {
            "Content-Type": "application/json"
          }
        });
        this.$store.commit("LOGIN");
        this.$router.push("/Landing");
        this.$store.commit("CHANGE_PAGE_NAME", "Landing");
      } catch (err) {
        console.error("ERROR LOGGING IN: ", err);
        console.log(err.response.status);
        const { status } = err.response;
        if (status === 401) {
          this.errorMessages = ["Incorrect Username or Password"];
        } else {
          this.errorMessages = [err.message]; // TODO: very user un-friendly error, but more useful than nothing :)
        }
      } finally {
        this.loading = false;
      }
    }
  }
};
</script>
