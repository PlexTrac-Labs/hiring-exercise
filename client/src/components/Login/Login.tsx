import React, { Dispatch, FormEvent, SetStateAction, useState } from "react";
import { Ctx } from "../../App";
import { Button, TextField } from "@material-ui/core";
import "./Login.scss";
import { setAccessToken } from "../../util/Token";

interface Props {
  setToken: Dispatch<SetStateAction<string>>;
}

export const LoginComponent: React.FC<Props> = props => {
  const ctx = React.useContext(Ctx);
  const [username, setUsername] = useState<string>("");
  const [password, setPassword] = useState<string>("");
  const [error, setError] = useState<string>("");

  const handleSubmit = async (e: FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    await ctx.authService
      .Login({
        username,
        password
      })
      .then(res => {
        props.setToken(res.auth_token);
        setAccessToken(res.auth_token);
        ctx.user = res.user;
      })
      .catch(_ => {
        setPassword("");
        setError("Login attempt failed. Please try again.");
      });
  };

  return (
    <div className="login-wrapper">
      <h1>Please Log In</h1>
      <form
        className="login-form"
        onSubmit={handleSubmit}
        onChange={() => setError("")}
      >
        <TextField
          label="Username"
          type="text"
          required={true}
          className="username-input"
          InputLabelProps={{
            className: "username-input"
          }}
          inputProps={{
            "data-testid": "username-input"
          }}
          value={username}
          onChange={e => setUsername(e.target.value)}
          error={false}
        />
        <TextField
          label="Password"
          type="password"
          required={true}
          className="password-input"
          InputLabelProps={{
            className: "password-input"
          }}
          inputProps={{
            "data-testid": "password-input"
          }}
          value={password}
          onChange={e => setPassword(e.target.value)}
        />
        <div className="error-container">
          {error && <p className="error-message">* {error}</p>}
        </div>
        <Button
          type="submit"
          className="login-btn"
          variant="outlined"
          data-testid="submit-btn"
        >
          Login
        </Button>
      </form>
    </div>
  );
};
