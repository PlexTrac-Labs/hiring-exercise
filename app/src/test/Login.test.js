import React from "react";
import ReactDOM from "react-dom";
import renderer from "react-test-renderer";
import Login from "../pages/Login";

//Check if login renders correctly
it("renders correctly", async () => {
  const tree = renderer.create(<Login />).toJSON();
  expect(tree).toMatchSnapshot();
});
