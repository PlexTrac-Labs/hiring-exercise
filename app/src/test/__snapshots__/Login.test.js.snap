// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="row align-items-center g-lg-5 py-5 col-md-10 mx-auto col-lg-5"
>
  <form
    className="pform-signin login"
    onSubmit={[Function]}
  >
    <div
      className="form-floating"
    >
      <input
        className="form-control"
        id="UserNameInput"
        name="name"
        required={true}
        type="text"
      />
      <label
        htmlFor="UserNameInput"
      >
        User Name
      </label>
    </div>
    <div
      className="form-floating"
    >
      <input
        className="form-control"
        id="passwordInput"
        name="password"
        required={true}
        type="text"
      />
      <label
        htmlFor="passwordInput"
      >
        Password
      </label>
    </div>
    <button
      className="w-100 btn btn-lg btn-primary"
      type="submit"
    >
      Log In
    </button>
  </form>
</div>
`;
